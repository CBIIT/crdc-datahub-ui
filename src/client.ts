import { ApolloClient, InMemoryCache, ApolloLink, HttpLink, DefaultOptions } from "@apollo/client";
import env from "./env";

const defaultOptions: DefaultOptions = {
  query: {
    fetchPolicy: "no-cache",
  },
};

const backendService = new HttpLink({
  uri: env.REACT_APP_BACKEND_API,
});

const mockService = new HttpLink({
  uri: "https://7a242248-52f7-476a-a60f-d64a2db3dd5b.mock.pstmn.io/graphql",
  headers: {
    "x-mock-match-request-body": "true",
  },
});

const cache = new InMemoryCache({
  typePolicies: {
    Submission: {
      keyFields: ["_id"],
      fields: {
        collaborators: {
          merge: false,
        },
      },
    },
    Collaborator: {
      keyFields: ["collaboratorID"],
    },
    PBACDefaults: {
      keyFields: ["role"],
    },
    Permission: {
      keyFields: false,
    },
    Notification: {
      keyFields: false,
    },
  },
});

const client = new ApolloClient({
  cache,
  defaultOptions,
  link: ApolloLink.split(
    (operation) => operation.getContext().clientName === "mockService",
    mockService,
    backendService
  ),
});

export default client;
