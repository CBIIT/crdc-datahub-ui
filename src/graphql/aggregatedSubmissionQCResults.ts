import gql from "graphql-tag";

// The base aggregatedQCResult model used for all aggregatedSubmissionQCResults queries
const BaseAggregatedQCResultFragment = gql`
  fragment BaseAggregatedQCResultFragment on aggregatedQCResult {
    code
    title
  }
`;

// The extended aggregatedQCResult model which includes all fields
const FullAggregatedQCResultFragment = gql`
  fragment AggregatedQCResultFragment on aggregatedQCResult {
    severity
    count
  }
`;

export const query = gql`
  query aggregatedSubmissionQCResults(
    $submissionID: ID!
    $severity: String
    $first: Int
    $offset: Int
    $orderBy: String
    $sortDirection: String
    $partial: Boolean = false
  ) {
    aggregatedSubmissionQCResults(
      submissionID: $submissionID
      severity: $severity
      first: $first
      offset: $offset
      orderBy: $orderBy
      sortDirection: $sortDirection
    ) {
      total
      results {
        ...BaseAggregatedQCResultFragment
        ...AggregatedQCResultFragment @skip(if: $partial)
      }
    }
  }
  ${FullAggregatedQCResultFragment}
  ${BaseAggregatedQCResultFragment}
`;

export type Input = {
  submissionID: string;
  severity?: string;
  first?: number;
  offset?: number;
  orderBy?: string;
  sortDirection?: string;
  partial?: boolean;
};

export type Response = {
  aggregatedSubmissionQCResults: ValidationResult<AggregatedQCResult>;
};
