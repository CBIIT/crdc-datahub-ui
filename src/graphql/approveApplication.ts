import gql from "graphql-tag";

export const mutation = gql`
  mutation approveApplication(
    $id: ID!
    $comment: String
    $wholeProgram: Boolean
    $institutions: [String]
    $pendingModelChange: Boolean
  ) {
    approveApplication(
      _id: $id
      wholeProgram: $wholeProgram
      comment: $comment
      institutions: $institutions
      pendingModelChange: $pendingModelChange
    ) {
      _id
    }
  }
`;

export type Input = {
  id: string;
  comment: string;
  wholeProgram: boolean;
  institutions: string[];
  pendingModelChange: boolean;
};

export type Response = {
  approveApplication: Pick<Application, "_id">;
};
