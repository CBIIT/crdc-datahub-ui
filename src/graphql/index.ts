// Submission Request
export { mutation as REVIEW_APP } from "./reviewApplication";
export type { Response as ReviewAppResp, Input as ReviewAppInput } from "./reviewApplication";

export { mutation as REOPEN_APP } from "./reopenApplication";
export type { Response as ReopenAppResp } from "./reopenApplication";

export { mutation as APPROVE_APP } from "./approveApplication";
export type { Input as ApproveAppInput, Response as ApproveAppResp } from "./approveApplication";

export { mutation as INQUIRE_APP } from "./inquireApplication";
export type { Response as InquireAppResp } from "./inquireApplication";

export { mutation as REJECT_APP } from "./rejectApplication";
export type { Response as RejectAppResp } from "./rejectApplication";

export { mutation as CANCEL_APP } from "./cancelApplication";
export type { Response as CancelAppResp, Input as CancelAppInput } from "./cancelApplication";

export { mutation as RESTORE_APP } from "./restoreApplication";
export type { Response as RestoreAppResp, Input as RestoreAppInput } from "./restoreApplication";

export { mutation as SAVE_APP } from "./saveApplication";
export type { Input as SaveAppInput, Response as SaveAppResp } from "./saveApplication";

export { mutation as SUBMIT_APP } from "./submitApplication";
export type { Response as SubmitAppResp } from "./submitApplication";

export { query as LAST_APP } from "./getMyLastApplication";
export type { Response as LastAppResp } from "./getMyLastApplication";

export { query as GET_APP } from "./getApplication";
export type { Response as GetAppResp } from "./getApplication";

export { mutation as UPDATE_MY_USER } from "./updateMyUser";
export type { Input as UpdateMyUserInput, Response as UpdateMyUserResp } from "./updateMyUser";

export { query as LIST_INSTITUTIONS } from "./listInstitutions";
export type { Response as ListInstitutionsResp } from "./listInstitutions";

export { query as RETRIEVE_CDEs } from "./retrieveCDEs";
export type { Response as RetrieveCDEsResp, Input as RetrieveCDEsInput } from "./retrieveCDEs";

export { query as LIST_APPLICATIONS } from "./listApplications";
export type {
  Response as ListApplicationsResp,
  Input as ListApplicationsInput,
} from "./listApplications";

// Data Submissions
export { mutation as CREATE_SUBMISSION } from "./createSubmission";
export type {
  Input as CreateSubmissionInput,
  Response as CreateSubmissionResp,
} from "./createSubmission";

export { query as GET_SUBMISSION } from "./getSubmission";
export type { Input as GetSubmissionInput, Response as GetSubmissionResp } from "./getSubmission";

export { query as LIST_SUBMISSIONS } from "./listSubmissions";
export type {
  Input as ListSubmissionsInput,
  Response as ListSubmissionsResp,
} from "./listSubmissions";

export { mutation as SUBMISSION_ACTION } from "./submissionAction";
export type { Response as SubmissionActionResp } from "./submissionAction";

export { mutation as CREATE_BATCH } from "./createBatch";
export type { Input as CreateBatchInput, Response as CreateBatchResp } from "./createBatch";

export { mutation as UPDATE_BATCH } from "./updateBatch";
export type { Response as UpdateBatchResp } from "./updateBatch";

export { query as LIST_BATCHES } from "./listBatches";
export type { Input as ListBatchesInput, Response as ListBatchesResp } from "./listBatches";

export { query as SUBMISSION_QC_RESULTS } from "./submissionQCResults";
export type {
  Input as SubmissionQCResultsInput,
  Response as SubmissionQCResultsResp,
} from "./submissionQCResults";

export { query as AGGREGATED_SUBMISSION_QC_RESULTS } from "./aggregatedSubmissionQCResults";
export type {
  Input as AggregatedSubmissionQCResultsInput,
  Response as AggregatedSubmissionQCResultsResp,
} from "./aggregatedSubmissionQCResults";

export { query as SUBMISSION_CROSS_VALIDATION_RESULTS } from "./submissionCrossValidationResults";
export type {
  Input as CrossValidationResultsInput,
  Response as CrossValidationResultsResp,
} from "./submissionCrossValidationResults";

export { mutation as VALIDATE_SUBMISSION } from "./validateSubmission";
export type {
  Input as ValidateSubmissionInput,
  Response as ValidateSubmissionResp,
} from "./validateSubmission";

export { query as LIST_NODE_TYPES } from "./listSubmissionNodeTypes";
export type {
  Input as ListNodeTypesInput,
  Response as ListNodeTypesResp,
} from "./listSubmissionNodeTypes";

export { query as GET_NODE_DETAIL } from "./getNodeDetail";
export type { Input as GetNodeDetailInput, Response as GetNodeDetailResp } from "./getNodeDetail";

export { query as GET_RELATED_NODES } from "./getRelatedNodes";
export type {
  Input as GetRelatedNodesInput,
  Response as GetRelatedNodesResp,
  PropertiesOnlyResponse as GetRelatedNodesRespPropsOnly,
} from "./getRelatedNodes";

export { query as GET_SUBMISSION_NODES } from "./getSubmissionNodes";
export type {
  Input as GetSubmissionNodesInput,
  Response as GetSubmissionNodesResp,
} from "./getSubmissionNodes";

export { query as RETRIEVE_RELEASED_DATA } from "./retrieveReleasedDataByID";
export type {
  Input as RetrieveReleasedDataInput,
  Response as RetrieveReleasedDataResp,
} from "./retrieveReleasedDataByID";

export { query as SUBMISSION_STATS } from "./submissionStats";
export type {
  Input as SubmissionStatsInput,
  Response as SubmissionStatsResp,
} from "./submissionStats";

export { mutation as DELETE_DATA_RECORDS } from "./deleteDataRecords";
export type {
  Input as DeleteDataRecordsInput,
  Response as DeleteDataRecordsResp,
} from "./deleteDataRecords";

export { query as RETRIEVE_CLI_CONFIG } from "./retrieveCLIConfig";
export type { Response as RetrieveCLIConfigResp } from "./retrieveCLIConfig";

export { query as LIST_POTENTIAL_COLLABORATORS } from "./listPotentialCollaborators";
export type {
  Input as ListPotentialCollaboratorsInput,
  Response as ListPotentialCollaboratorsResp,
} from "./listPotentialCollaborators";

export { mutation as EDIT_SUBMISSION_COLLABORATORS } from "./editSubmissionCollaborators";
export type {
  Input as EditSubmissionCollaboratorsInput,
  Response as EditSubmissionCollaboratorsResp,
} from "./editSubmissionCollaborators";

// User Profile
export { query as GET_MY_USER } from "./getMyUser";
export type { Response as GetMyUserResp } from "./getMyUser";

export { query as GET_USER } from "./getUser";
export type { Input as GetUserInput, Response as GetUserResp } from "./getUser";

export { query as LIST_USERS } from "./listUsers";
export type { Response as ListUsersResp } from "./listUsers";

export { mutation as EDIT_USER } from "./editUser";
export type { Input as EditUserInput, Response as EditUserResp } from "./editUser";

export { mutation as REQUEST_ACCESS } from "./requestAccess";
export type { Input as RequestAccessInput, Response as RequestAccessResp } from "./requestAccess";

export { query as RETRIEVE_PBAC_DEFAULTS } from "./retrievePBACDefaults";
export type {
  Input as RetrievePBACDefaultsInput,
  Response as RetrievePBACDefaultsResp,
} from "./retrievePBACDefaults";

// Organizations
export { query as LIST_ORGS } from "./listOrganizations";
export type { Response as ListOrgsResp } from "./listOrganizations";

export { query as GET_ORG } from "./getOrganization";
export type { Response as GetOrgResp } from "./getOrganization";

export { mutation as EDIT_ORG } from "./editOrganization";
export type { Input as EditOrgInput, Response as EditOrgResp } from "./editOrganization";

export { query as LIST_ACTIVE_DCPS } from "./listActiveDCPs";
export type { Response as ListActiveDCPsResp } from "./listActiveDCPs";

export { query as LIST_APPROVED_STUDIES } from "./listApprovedStudies";
export type {
  Input as ListApprovedStudiesInput,
  Response as ListApprovedStudiesResp,
} from "./listApprovedStudies";

export { mutation as CREATE_APPROVED_STUDY } from "./createApprovedStudy";
export type {
  Input as CreateApprovedStudyInput,
  Response as CreateApprovedStudyResp,
} from "./createApprovedStudy";

export { mutation as UPDATE_APPROVED_STUDY } from "./updateApprovedStudy";
export type {
  Input as UpdateApprovedStudyInput,
  Response as UpdateApprovedStudyResp,
} from "./updateApprovedStudy";

export { query as GET_APPROVED_STUDY } from "./getApprovedStudy";
export type {
  Input as GetApprovedStudyInput,
  Response as GetApprovedStudyResp,
} from "./getApprovedStudy";

export { mutation as CREATE_ORG } from "./createOrganization";
export type { Input as CreateOrgInput, Response as CreateOrgResp } from "./createOrganization";

// Misc.
export { mutation as GRANT_TOKEN } from "./grantToken";
export type { Response as GrantTokenResp } from "./grantToken";

// Operation Dashboard
export { query as GET_DASHBOARD_URL } from "./getDashboardURL";
export type {
  Input as GetDashboardURLInput,
  Response as GetDashboardURLResp,
} from "./getDashboardURL";
