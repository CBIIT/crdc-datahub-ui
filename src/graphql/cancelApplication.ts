import { TypedDocumentNode } from "@apollo/client";
import gql from "graphql-tag";

export const mutation: TypedDocumentNode<Response, Input> = gql`
  mutation cancelApplication($_id: ID!) {
    cancelApplication: deleteApplication(_id: $_id) {
      _id
    }
  }
`;

export type Input = {
  /**
   * ID of the application to cancel
   */
  _id: string;
};

export type Response = {
  cancelApplication: Pick<Application, "_id">;
};
