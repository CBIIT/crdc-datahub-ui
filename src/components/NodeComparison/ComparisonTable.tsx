import {
  Alert,
  Skeleton,
  styled,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
} from "@mui/material";
import { FC, memo, useMemo } from "react";
import { RetrieveReleasedDataResp } from "../../graphql";
import { safeParse } from "../../utils";
import Repeater from "../Repeater";

const StyledAlert = styled(Alert)({
  marginTop: "12px",
});

const StyledTableContainer = styled(TableContainer)({
  marginTop: "12px",
  paddingBottom: "6px",
  overflowX: "auto",
  width: "100%",
});

const StyledTable = styled(Table)({
  backgroundColor: "#fff",
  borderCollapse: "separate",
  borderRadius: "8px",
  tableLayout: "auto",
  // NOTE: The first element in this thead is a TD, not a TH
  "& thead tr:first-of-type td:first-of-type": {
    borderTopLeftRadius: "8px",
  },
  "& thead tr:first-of-type th:last-child": {
    borderTopRightRadius: "8px",
  },
  "& tbody tr:last-child td:first-of-type": {
    borderBottomLeftRadius: "8px",
  },
  "& tbody tr:last-child td:last-child": {
    borderBottomRightRadius: "8px",
  },
});

const StyledTableHead = styled(TableHead)({
  "& .MuiTableRow-root th": {
    fontWeight: 700,
  },
});

const StyledTableCell = styled(TableCell)({
  color: "#083A50",
  whiteSpace: "nowrap",
  overflow: "hidden",
  border: "0.5px solid #a7a7a7",
  padding: "8px 15px",
});

/**
 * The number of placeholder columns to display in the loading state
 */
const PLACEHOLDER_NUM_COLS = 5;

/**
 * The width of the new/existing column
 */
const BLANK_COL_WIDTH = 55;

export type ComparisonTableProps = {
  /**
   * The new node to be compared
   */
  newNode: RetrieveReleasedDataResp["retrieveReleasedDataByID"][number];
  /**
   * The existing node to be compared
   */
  existingNode: RetrieveReleasedDataResp["retrieveReleasedDataByID"][number];
  /**
   * A boolean indicating whether the data is in a loading state
   */
  loading: boolean;
};

/**
 * Builds a dynamic table to compare the dataset between two data records
 * Handles loading state
 *
 * @returns The NodeComparisonTable component
 */
const ComparisonTable: FC<ComparisonTableProps> = ({ newNode, existingNode, loading }) => {
  const newProps = useMemo<Record<string, string | number>>(
    () => safeParse(newNode?.props),
    [newNode]
  );

  const existingProps = useMemo<Record<string, string | number>>(
    () => safeParse(existingNode?.props),
    [existingNode]
  );

  const allPropertyNames = useMemo<string[]>(() => {
    const newKeys = Object.keys(newProps);
    const existingKeys = Object.keys(existingProps);
    return [...new Set([...newKeys, ...existingKeys])];
  }, [newProps, existingProps]);

  if (!loading && !allPropertyNames.length) {
    return (
      <StyledAlert severity="warning" data-testid="node-comparison-error">
        Unable to show comparison of data.
      </StyledAlert>
    );
  }

  return (
    <StyledTableContainer>
      <StyledTable data-testid="node-comparison-table" size="small">
        <StyledTableHead>
          <TableRow>
            <StyledTableCell component="td" width={BLANK_COL_WIDTH} />
            {loading ? (
              <TableHeaderSkeleton />
            ) : (
              allPropertyNames.map((property) => (
                <StyledTableCell key={property}>{property}</StyledTableCell>
              ))
            )}
          </TableRow>
        </StyledTableHead>
        <TableBody>
          {loading ? (
            <TableBodySkeleton />
          ) : (
            <>
              <TableRow data-testid="node-comparison-table-existing">
                <StyledTableCell width={BLANK_COL_WIDTH}>Existing</StyledTableCell>
                {allPropertyNames.map((property) => (
                  <StyledTableCell key={property}>
                    {existingProps?.[property] || ""}
                  </StyledTableCell>
                ))}
              </TableRow>
              <TableRow data-testid="node-comparison-table-new">
                <StyledTableCell width={BLANK_COL_WIDTH}>New</StyledTableCell>
                {allPropertyNames.map((property) => (
                  <StyledTableCell key={property}>{newProps?.[property] || ""}</StyledTableCell>
                ))}
              </TableRow>
            </>
          )}
        </TableBody>
      </StyledTable>
    </StyledTableContainer>
  );
};

/**
 * A skeleton table header to display while the data is loading
 *
 * @returns The TableHeaderSkeleton component
 */
const TableHeaderSkeleton: FC = () => (
  <Repeater count={PLACEHOLDER_NUM_COLS} keyPrefix="loader-skeleton-header">
    <StyledTableCell component="td" data-testid="node-comparison-table-header-skeleton">
      <Skeleton variant="text" width="100%" height={24} />
    </StyledTableCell>
  </Repeater>
);

/**
 * A skeleton table body to display while the data is loading
 *
 * @returns The TableBodySkeleton component
 */
const TableBodySkeleton: FC = () => (
  <>
    <TableRow>
      <StyledTableCell width={BLANK_COL_WIDTH}>Existing</StyledTableCell>
      <Repeater count={PLACEHOLDER_NUM_COLS} keyPrefix="loader-skeleton-existing">
        <StyledTableCell>
          <Skeleton variant="text" width="100%" height={24} />
        </StyledTableCell>
      </Repeater>
    </TableRow>
    <TableRow>
      <StyledTableCell width={BLANK_COL_WIDTH}>New</StyledTableCell>
      <Repeater count={PLACEHOLDER_NUM_COLS} keyPrefix="loader-skeleton-new">
        <StyledTableCell>
          <Skeleton variant="text" width="100%" height={24} />
        </StyledTableCell>
      </Repeater>
    </TableRow>
  </>
);

export default memo<ComparisonTableProps>(ComparisonTable);
