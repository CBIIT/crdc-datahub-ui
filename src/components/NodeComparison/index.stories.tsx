import { MockedResponse } from "@apollo/client/testing";
import type { Meta, StoryObj } from "@storybook/react";

import {
  RETRIEVE_RELEASED_DATA,
  RetrieveReleasedDataInput,
  RetrieveReleasedDataResp,
} from "../../graphql";

import NodeComparison from "./index";

const nominalMock: MockedResponse<RetrieveReleasedDataResp, RetrieveReleasedDataInput> = {
  request: {
    query: RETRIEVE_RELEASED_DATA,
  },
  variableMatcher: () => true,
  result: {
    data: {
      retrieveReleasedDataByID: [
        // New node
        {
          nodeType: "participant",
          nodeID: "participant-123",
          props: JSON.stringify({
            col1: "xyz",
            equal_val: "we're equal",
            col2: "abc",
            unchanged_val: "",
            identical_delete: "<delete>",
            mock_col: "foo",
            delete_val: "<delete>",
          }),
        },
        // Existing node
        {
          nodeType: "participant",
          nodeID: "participant-123",
          props: JSON.stringify({
            col1: "123",
            equal_val: "we're equal",
            col2: "xyz",
            unchanged_val: "wont change",
            identical_delete: "<delete>",
            mock_col: "abc",
            delete_val: "some val to be deleted",
          }),
        },
      ],
    },
  },
};

const errorMock: MockedResponse<RetrieveReleasedDataResp, RetrieveReleasedDataInput> = {
  request: {
    query: RETRIEVE_RELEASED_DATA,
  },
  variableMatcher: () => true,
  error: new Error("GraphQL error"),
};

const loadingMock: MockedResponse<RetrieveReleasedDataResp, RetrieveReleasedDataInput> = {
  request: {
    query: RETRIEVE_RELEASED_DATA,
  },
  variableMatcher: () => true,
  result: null,
  delay: Infinity,
};

// --- Storybook Meta ---
const meta: Meta<typeof NodeComparison> = {
  title: "Miscellaneous / Node Comparison",
  component: NodeComparison,
  tags: ["autodocs"],
  args: {
    submissionID: "mock-submission-id",
    nodeType: "mock-node",
    submittedID: "mock-node-id",
  },
} satisfies Meta<typeof NodeComparison>;

export default meta;

type Story = StoryObj<typeof NodeComparison>;

/**
 * This story represents the default state of the component with mock data.
 */
export const DefaultState: Story = {
  name: "Default",
  parameters: {
    apolloClient: {
      mocks: [nominalMock],
    },
  },
};

/**
 * This story simulates a loading state for the component.
 */
export const LoadingState: Story = {
  name: "Loading",
  parameters: {
    apolloClient: {
      mocks: [loadingMock],
    },
  },
};

/**
 * This story simulates handling of an API error.
 */
export const ErrorState: Story = {
  name: "Error",
  parameters: {
    apolloClient: {
      mocks: [errorMock],
    },
  },
};
