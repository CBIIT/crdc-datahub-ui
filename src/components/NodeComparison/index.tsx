import { useQuery } from "@apollo/client";
import { Box, styled, Typography } from "@mui/material";
import { FC, memo, useMemo } from "react";
import { useSnackbar } from "notistack";
import {
  RETRIEVE_RELEASED_DATA,
  RetrieveReleasedDataInput,
  RetrieveReleasedDataResp,
} from "../../graphql";
import { Logger } from "../../utils";
import ComparisonTable from "./ComparisonTable";

const StyledBox = styled(Box)({
  padding: "10px",
});

const StyledHeadingTypography = styled(Typography)({
  color: "#595959",
  fontSize: "14px",
});

export type NodeComparisonProps = {
  /**
   * The identifier for the Data Submission which contains the node
   *
   * @example "submission_10290"
   */
  submissionID: string;
  /**
   * The distinct type of the node to be compared
   *
   * @example "participant"
   */
  nodeType: string;
  /**
   * The identifier for the node
   *
   * @example "participant_10290"
   */
  submittedID: string;
};

/**
 * A component that renders the existing/new data comparison table of two nodes
 *
 * @returns The DataComparison component
 */
const NodeComparison: FC<NodeComparisonProps> = ({ submissionID, nodeType, submittedID }) => {
  const { enqueueSnackbar } = useSnackbar();
  const { data, loading } = useQuery<RetrieveReleasedDataResp, RetrieveReleasedDataInput>(
    RETRIEVE_RELEASED_DATA,
    {
      variables: {
        submissionId: submissionID,
        nodeType,
        nodeId: submittedID,
      },
      fetchPolicy: "cache-and-network",
      onError: (error) => {
        Logger.error("NodeComparison API error received", error);
        enqueueSnackbar("Oops! Unable to retrieve the data record comparison", {
          variant: "error",
        });
      },
    }
  );

  const isLoading = useMemo<boolean>(() => loading && !data, [loading]);

  const [newNode, existingNode] = useMemo<
    [
      RetrieveReleasedDataResp["retrieveReleasedDataByID"][number],
      RetrieveReleasedDataResp["retrieveReleasedDataByID"][number],
    ]
  >(() => {
    if (isLoading) {
      return [null, null];
    }
    if (data?.retrieveReleasedDataByID?.length !== 2) {
      Logger.error("NodeComparison API did not return exactly 2 nodes", data);
      return [null, null];
    }

    return [data.retrieveReleasedDataByID[0], data.retrieveReleasedDataByID[1]];
  }, [data, isLoading]);

  return (
    <StyledBox>
      <StyledHeadingTypography variant="body1" data-testid="node-comparison-header">
        A record with this ID already exists. Review the existing and newly submitted data to decide
        whether to update the current record.
      </StyledHeadingTypography>
      <ComparisonTable newNode={newNode} existingNode={existingNode} loading={isLoading} />
    </StyledBox>
  );
};

export default memo<NodeComparisonProps>(NodeComparison);
