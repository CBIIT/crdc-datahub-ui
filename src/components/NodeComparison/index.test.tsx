import { act, render, waitFor } from "@testing-library/react";
import { axe } from "jest-axe";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import { FC } from "react";
import { GraphQLError } from "graphql";
import {
  RETRIEVE_RELEASED_DATA,
  RetrieveReleasedDataInput,
  RetrieveReleasedDataResp,
} from "../../graphql";
import NodeComparison from "./index";

const mockTableRender = jest.fn().mockImplementation(() => <div>MOCK COMPARISON TABLE</div>);
jest.mock("./ComparisonTable", () => ({
  __esModule: true,
  default: (...p) => mockTableRender(...p),
}));

type MockParentProps = {
  mocks?: MockedResponse[];
  children: React.ReactNode;
};

const MockParent: FC<MockParentProps> = ({ mocks, children }) => (
  <MockedProvider mocks={mocks} addTypename={false}>
    {children}
  </MockedProvider>
);

describe("Accessibility", () => {
  it("should have no violations", async () => {
    const mock: MockedResponse<RetrieveReleasedDataResp, RetrieveReleasedDataInput> = {
      request: {
        query: RETRIEVE_RELEASED_DATA,
      },
      variableMatcher: () => true,
      result: {
        data: {
          retrieveReleasedDataByID: [],
        },
      },
    };

    const { container } = render(<NodeComparison submissionID="" nodeType="" submittedID="" />, {
      wrapper: ({ children }) => <MockParent mocks={[mock]}>{children}</MockParent>,
    });

    await act(async () => {
      expect(await axe(container)).toHaveNoViolations();
    });
  });
});

describe("Basic Functionality", () => {
  it("should show an error message when nodes cannot be fetch (Network)", async () => {
    const mock: MockedResponse<RetrieveReleasedDataResp, RetrieveReleasedDataInput> = {
      request: {
        query: RETRIEVE_RELEASED_DATA,
      },
      variableMatcher: () => true,
      error: new Error("GraphQL error"),
    };

    render(<NodeComparison submissionID="" nodeType="" submittedID="" />, {
      wrapper: ({ children }) => <MockParent mocks={[mock]}>{children}</MockParent>,
    });

    await waitFor(() => {
      expect(global.mockEnqueue).toHaveBeenCalledWith(
        "Oops! Unable to retrieve the data record comparison",
        { variant: "error" }
      );
    });
  });

  it("should show an error message when nodes cannot be fetch (GraphQL)", async () => {
    const mock: MockedResponse<RetrieveReleasedDataResp, RetrieveReleasedDataInput> = {
      request: {
        query: RETRIEVE_RELEASED_DATA,
      },
      variableMatcher: () => true,
      result: {
        errors: [new GraphQLError("GraphQL mock error")],
      },
    };

    render(<NodeComparison submissionID="" nodeType="" submittedID="" />, {
      wrapper: ({ children }) => <MockParent mocks={[mock]}>{children}</MockParent>,
    });

    await waitFor(() => {
      expect(global.mockEnqueue).toHaveBeenCalledWith(
        "Oops! Unable to retrieve the data record comparison",
        { variant: "error" }
      );
    });
  });

  it("should forward the node data to the ComparisonTable", async () => {
    const mock: MockedResponse<RetrieveReleasedDataResp, RetrieveReleasedDataInput> = {
      request: {
        query: RETRIEVE_RELEASED_DATA,
      },
      variableMatcher: () => true,
      result: {
        data: {
          retrieveReleasedDataByID: [
            {
              nodeType: "mock_node_type",
              nodeID: "mock_node_id",
              props: JSON.stringify({ mock_node_data_name: "foo", baz: 1 }),
            },
            {
              nodeType: "mock_node_type",
              nodeID: "mock_node_id",
              props: JSON.stringify({ mock_node_data_name: "bar", baz: 2 }),
            },
          ],
        },
      },
    };

    render(<NodeComparison submissionID="" nodeType="" submittedID="" />, {
      wrapper: ({ children }) => <MockParent mocks={[mock]}>{children}</MockParent>,
    });

    await waitFor(() => {
      // Calls twice because of the loading state
      expect(mockTableRender).toHaveBeenCalledWith(
        expect.objectContaining({
          newNode: expect.objectContaining({
            nodeType: "mock_node_type",
            nodeID: "mock_node_id",
            props: JSON.stringify({ mock_node_data_name: "foo", baz: 1 }),
          }),
          existingNode: expect.objectContaining({
            nodeType: "mock_node_type",
            nodeID: "mock_node_id",
            props: JSON.stringify({ mock_node_data_name: "bar", baz: 2 }),
          }),
          loading: false,
        }),
        {}
      );
    });
  });

  it("should forward null, null to the ComparisonTable if the API returns a non-nominal array length (1)", async () => {
    const mock: MockedResponse<RetrieveReleasedDataResp, RetrieveReleasedDataInput> = {
      request: {
        query: RETRIEVE_RELEASED_DATA,
      },
      variableMatcher: () => true,
      result: {
        data: {
          retrieveReleasedDataByID: [
            {
              nodeType: "mock_node_type",
              nodeID: "mock_node_id",
              props: JSON.stringify({ mock_node_data_name: "foo", baz: 1 }),
            },
          ],
        },
      },
    };

    render(<NodeComparison submissionID="" nodeType="" submittedID="" />, {
      wrapper: ({ children }) => <MockParent mocks={[mock]}>{children}</MockParent>,
    });

    await waitFor(() => {
      expect(mockTableRender).toHaveBeenCalledWith(
        expect.objectContaining({
          newNode: null,
          existingNode: null,
          loading: false,
        }),
        {}
      );
    });
  });

  it("should forward null, null to the ComparisonTable if the API returns a non-nominal array length (3)", async () => {
    const mock: MockedResponse<RetrieveReleasedDataResp, RetrieveReleasedDataInput> = {
      request: {
        query: RETRIEVE_RELEASED_DATA,
      },
      variableMatcher: () => true,
      result: {
        data: {
          retrieveReleasedDataByID: [
            {
              nodeType: "mock_node_type",
              nodeID: "mock_node_id",
              props: JSON.stringify({ mock_node_data_name: "foo", baz: 1 }),
            },
            {
              nodeType: "mock_node_type",
              nodeID: "mock_node_id",
              props: JSON.stringify({ mock_node_data_name: "foo", baz: 1 }),
            },
            {
              nodeType: "mock_node_type",
              nodeID: "mock_node_id",
              props: JSON.stringify({ mock_node_data_name: "foo", baz: 1 }),
            },
          ],
        },
      },
    };

    render(<NodeComparison submissionID="" nodeType="" submittedID="" />, {
      wrapper: ({ children }) => <MockParent mocks={[mock]}>{children}</MockParent>,
    });

    await waitFor(() => {
      expect(mockTableRender).toHaveBeenCalledWith(
        expect.objectContaining({
          newNode: null,
          existingNode: null,
          loading: false,
        }),
        {}
      );
    });
  });
});

describe("Implementation Requirements", () => {
  it("should have a header describing the comparison", () => {
    const mock: MockedResponse<RetrieveReleasedDataResp, RetrieveReleasedDataInput> = {
      request: {
        query: RETRIEVE_RELEASED_DATA,
      },
      variableMatcher: () => true,
      result: {
        data: {
          retrieveReleasedDataByID: [],
        },
      },
    };

    const { getByTestId } = render(<NodeComparison submissionID="" nodeType="" submittedID="" />, {
      wrapper: ({ children }) => <MockParent mocks={[mock]}>{children}</MockParent>,
    });

    expect(getByTestId("node-comparison-header")).toBeInTheDocument();
    expect(getByTestId("node-comparison-header")).toHaveTextContent(
      /A record with this ID already exists. Review the existing and newly submitted data to decide whether to update the current record./
    );
  });
});

describe("Snapshots", () => {
  it("should match the nominal state snapshot", async () => {
    const mockMatcher = jest.fn().mockReturnValue(true);
    const mock: MockedResponse<RetrieveReleasedDataResp, RetrieveReleasedDataInput> = {
      request: {
        query: RETRIEVE_RELEASED_DATA,
      },
      variableMatcher: mockMatcher,
      result: {
        data: {
          retrieveReleasedDataByID: [],
        },
      },
    };

    const { container } = render(<NodeComparison submissionID="" nodeType="" submittedID="" />, {
      wrapper: ({ children }) => <MockParent mocks={[mock]}>{children}</MockParent>,
    });

    await waitFor(() => {
      expect(mockMatcher).toHaveBeenCalled(); // wait for the state to normalize
    });

    expect(container).toMatchSnapshot();
  });
});
