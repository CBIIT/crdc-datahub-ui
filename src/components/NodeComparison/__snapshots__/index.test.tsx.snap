// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots should match the nominal state snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-17gs3ot"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-1r7ezte-MuiTypography-root"
      data-testid="node-comparison-header"
    >
      A record with this ID already exists. Review the existing and newly submitted data to decide whether to update the current record.
    </p>
  </div>
</div>
`;
